Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import json
+import datetime
+import matplotlib.pyplot as plt
+import pandas as pd
+
+
+# In[7]:
+
+
+class Metric():
+    
+    def __init__(self, path, since=None, until=None, source_code_exclude_list=None):
+
+        self.since = since
+        self.until = until
+        self.source_code_exclude_list = source_code_exclude_list
+        self.repo_urls = set()
+        
+        clean_commit_list = list()
+        clean_issue_list = list()
+        clean_pr_list = list()
+        
+        with open(path, 'r') as raw_data:
+            for line in raw_data:
+                line = json.loads(line)
+                self.repo_urls.add(line['origin'])
+                clean_line = dict()
+                if line['category'] == "commit":
+                    clean_line = self._clean_commit(line)
+
+                    if self._is_source_code(clean_line):
+                        clean_commit_list.append(clean_line)
+                    
+
+                elif line['category'] == "issue":
+                    if "pull_request" not in line['data']:
+                        clean_line = self._clean_issue(line)
+                        clean_issue_list.append(clean_line)
+                    else: continue
+                    
+
+                elif line['category'] == "pull_request":
+                    clean_line = self._clean_pr(line)
+                    clean_pr_list.append(clean_line)
+                        
+                        
+                self.clean_commit_df = pd.DataFrame(clean_commit_list)
+                self.clean_issue_df = pd.DataFrame(clean_issue_list)
+                self.clean_pr_df = pd.DataFrame(clean_pr_list)
+                
+            self.clean_dict = {
```

Comment:
What is this for? In general, if a class needs commits, it should read a JSON file with commits...

Reply:
A user might append all kinds of data in a single JSON file, the way I did in my microtasks:
```python
  for repo, repo_url in zip(repos_used, repo_urls):

      !perceval git --json-line $repo_url >> ../$file_name

      !perceval github -t $auth_token --json-line --sleep-for-rate --category pull_request $owner $repo >> ../$file_name

      !perceval github -t $auth_token --json-line --sleep-for-rate --category issue $owner $repo >> ../$file_name
```

Hence, in the root class, I organized everything in the file based on the category: issue, pr or commit and placed them in their corresponding dataframes. 
Now, if a metric class needs to work only with commits, it could do something like this:

```python
self.metric_df = self.clean_dict["commit"]
```
And then complete the analysis and calculations based on `self.metric_df`

Reply:
OK, now I think I start to understand your rationale. But remember that in this case, we're not writing generic scripts, that can deal with as much cases as reasonably possible. On the contrary, we should produce implementations as clean as possible, maybe at the cost of being specific to accepting only some inputs. In this case, we can assume that a metric that is going to deal with commits will be fed with only commits. Of course, we will need to be careful in documenting this restriction, but that will make the class much easier.